plugin/pkg/managedseed/validator/admission.go

```diff
@@ -0,0 +1,463 @@
+
+	secretBindingInformer := f.Core().InternalVersion().SecretBindings()
+	v.secretBindingLister = secretBindingInformer.Lister()
+
+	readyFuncs = append(readyFuncs, shootInformer.Informer().HasSynced, secretBindingInformer.Informer().HasSynced)
+}
+
+// SetKubeInformerFactory gets Lister from SharedInformerFactory.
+func (v *ValidateManagedSeed) SetKubeInformerFactory(f kubeinformers.SharedInformerFactory) {
+	secretInformer := f.Core().V1().Secrets()
+	v.secretLister = secretInformer.Lister()
+
+	readyFuncs = append(readyFuncs, secretInformer.Informer().HasSynced)
+}
+
+// SetInternalCoreClientset sets the garden core clientset.
+func (v *ValidateManagedSeed) SetInternalCoreClientset(c coreclientset.Interface) {
+	v.coreClient = c
+}
+
+// SetSeedManagementClientset sets the garden seedmanagement clientset.
+func (v *ValidateManagedSeed) SetSeedManagementClientset(c seedmanagementclientset.Interface) {
+	v.seedManagementClient = c
+}
+
+// ValidateInitialization checks whether the plugin was correctly initialized.
+func (v *ValidateManagedSeed) ValidateInitialization() error {
+	if v.shootLister == nil {
+		return errors.New("missing shoot lister")
+	}
+	if v.secretBindingLister == nil {
+		return errors.New("missing secret binding lister")
+	}
+	if v.secretLister == nil {
+		return errors.New("missing secret lister")
+	}
+	if v.coreClient == nil {
+		return errors.New("missing garden core client")
+	}
+	if v.seedManagementClient == nil {
+		return errors.New("missing garden seedmanagement client")
+	}
+	return nil
+}
+
+var _ admission.MutationInterface = &ValidateManagedSeed{}
+
+// Admit validates and if appropriate mutates the given managed seed against the shoot that it references.
+func (v *ValidateManagedSeed) Admit(ctx context.Context, a admission.Attributes, o admission.ObjectInterfaces) error {
+	// Wait until the caches have been synced
+	if v.readyFunc == nil {
+		v.AssignReadyFunc(func() bool {
+			for _, readyFunc := range readyFuncs {
+				if !readyFunc() {
+					return false
+				}
+			}
+			return true
+		})
+	}
+	if !v.WaitForReady() {
+		return admission.NewForbidden(a, errors.New("not yet ready to handle request"))
+	}
+
+	// Ignore all kinds other than ManagedSeed
+	if a.GetKind().GroupKind() != seedmanagement.Kind("ManagedSeed") {
+		return nil
+	}
+
+	// Ignore updates to status or other subresources
+	if a.GetSubresource() != "" {
+		return nil
+	}
+
+	// Convert object to ManagedSeed
+	managedSeed, ok := a.GetObject().(*seedmanagement.ManagedSeed)
+	if !ok {
+		return apierrors.NewBadRequest("could not convert object to ManagedSeed")
+	}
+
+	var allErrs field.ErrorList
+	gk := schema.GroupKind{Group: seedmanagement.GroupName, Kind: "ManagedSeed"}
+
+	// Ensure shoot name is specified
+	shootNamePath := field.NewPath("spec", "shoot", "name")
+	if managedSeed.Spec.Shoot.Name == "" {
+		return apierrors.NewInvalid(gk, managedSeed.Name, append(allErrs, field.Required(shootNamePath, "shoot name is required")))
+	}
```

Comment:
I am afraid there is a need. You could try it yourself - remove the lines above and then create a `ManagedSeed` with an empty shoot name to see what happens.

Spoiler: Instead of a neat error message from validation, you would get an ugly one saying that a shoot named "" could not be found. The reason is that the admission plugin kicks in before validation, and it needs to read the shoot.

In fact, the check in validation is the one that could be removed. However, I decided to keep it there as well because it's quite an obvious thing to do and it could disturb other people (including reviewers) that it't not there.

Reply:
> The reason is that the admission plugin kicks in before validation, and it needs to read the shoot.

Ok, I didn't have this in mind. But you're right of course.
Thanks for clarifying.